File: vision/image.py
from dataclasses import dataclass
from pathlib import Path
from typing import TypeVar

import numpy as np
import numpy.typing as npt
from PIL import Image as PillowImage

TImage = TypeVar("TImage", bound="Image")


@dataclass
class Image:
    """A container class for performing various image related manipulations."""

    data: npt.NDArray[np.uint8]

    def blur(self: TImage, kernel_size: int, sigma: float) -> TImage:
        """Perform a Gaussian blurring of the image.

        Keyword arguments:
        kernel_size -- the size of the Gaussian kernel
        sigma -- the standard deviation for the Gaussian kernel
        """

        kernel = compute_gaussian_kernel(kernel_size, sigma)

        output = np.zeros_like(self.data)
        self.data = np.pad(self.data, pad_width=int((kernel_size - 1) / 2), mode="edge")
        for y in range(self.data.shape[1] - kernel_size + 1):
            for x in range(self.data.shape[0] - kernel_size + 1):
                output[x, y] = np.sum(
                    kernel
                    * self.data[
                        x : x + kernel_size,
                        y : y + kernel_size,
                    ]
                )

        self.data = output.astype(np.uint8)
        return self

    def segment(self: TImage, threshold: int = 240) -> TImage:
        """Perform a basic image segmentation of the image.

        The image is segmented into two regions, where the foreground is defined as the parts of the image with lower intensities.
        The foreground elements are mapped to 1 while the background elements are mapped to 0.

        Keyword arguments:
        threshold -- intensities smaller than this value will be included in the foreground region of the segmentation
        """

        output = np.zeros_like(self.data)
        output[self.data < threshold] = 1
        output[self.data >= threshold] = 0
        return type(self)(data=output)

    @property
    def center_of_gravity(self) -> tuple[float, float]:
        """Compute the center of gravity of the foreground region of an image."""

        image_segmented = self.segment()
        return tuple(np.average(np.nonzero(image_segmented.data), axis=1))  # type: ignore

    @property
    def principal_axes(self) -> npt.NDArray[np.float64]:
        """Compute the principal axes of the foreground region of an image.

        The eigenvectors constituting the principal axes are returned in descending order
        based on their corresponding eigenvalues.
        """

        image_segmented = self.segment()
        region_coordinates = np.array(np.nonzero(image_segmented.data)).T
        num_points = np.sum(image_segmented.data)
        mean_vector = 1 / num_points * np.sum(region_coordinates, axis=0)

        covariance_matrix = (
            (1 / (num_points - 1))
            * (region_coordinates - mean_vector).T
            @ (region_coordinates - mean_vector)
        )

        eigenvalues, eigenvectors = np.linalg.eig(covariance_matrix)

        idx = eigenvalues.argsort()[::-1]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]

        return eigenvectors  # type: ignore

    def get_image_with_normalized_foreground_region(self: TImage) -> TImage:
        """Return a boundary image with the foreground transformed such that it has a normalized translation and rotation."""

        image_segmented = self.segment()
        region_coordinates = np.array(np.nonzero(image_segmented.data)).T
        coordinates_transformed = (
            (region_coordinates - self.center_of_gravity) @ self.principal_axes
            + tuple(dim / 2 for dim in self.data.shape)
        ).astype(np.int64)

        output = np.zeros_like(image_segmented.data)
        output[coordinates_transformed[:, 0], coordinates_transformed[:, 1]] = 255

        return type(self)(data=output)


def load(path: Path, shape: tuple[int, int]) -> Image:
    """Load an image stored in 8-bit binary raw format (.bin) with a known shape."""
    data = np.fromfile(path, dtype=np.uint8).reshape(shape)
    return Image(data=data)


def save(path: Path, image: Image) -> None:
    """Save an image to disk.

    The image format is inferred from the specified path.
    """
    image_pillow = PillowImage.fromarray(image.data)
    image_pillow.save(path)


def compute_gaussian_kernel(kernel_size: int, sigma: float) -> npt.NDArray[np.float64]:
    """Compute the normalized kernel used for Gaussian blurring.

    Keyword arguments:
    kernel_size -- the size of the Gaussian kernel
    sigma -- the standard deviation for the Gaussian kernel
    """
    kernel_range = np.arange(kernel_size)
    s, t = np.meshgrid(kernel_range, kernel_range)
    center = (kernel_size - 1) / 2
    r = np.sqrt((s - center) ** 2 + (t - center) ** 2)

    kernel = np.exp(-0.5 * (r / sigma) ** 2).astype(np.float64)
    scale = 1.0 / np.sum(kernel)

    return scale * kernel

File: vision/cpp/Image.cpp
#include "Image.hpp"

#include <iostream>
#include <numbers>

namespace vision {
namespace image {

Image Image::Blur(size_t kernel_size, double sigma) {
  auto gaussian_kernel = [](size_t kernel_size, double sigma) {
    Eigen::MatrixXdRowMajor kernel(kernel_size, kernel_size);

    double center = (kernel_size - 1) / 2;
    for (size_t s = 0, num_rows = kernel.rows(), num_cols = kernel.cols();
         s < num_rows; ++s) {
      for (size_t t = 0; t < num_cols; ++t) {
        kernel(s, t) = exp(-0.5 * (pow((s - center) / sigma, 2.0) +
                                   pow((t - center) / sigma, 2.0))) /
                       (2 * std::numbers::pi * sigma * sigma);
      }
    }

    double sum = kernel.sum();
    for (size_t s = 0, num_rows = kernel.rows(), num_cols = kernel.cols();
         s < num_rows; ++s) {
      for (size_t t = 0; t < num_cols; ++t) {
        kernel(s, t) /= sum;
      }
    }

    return kernel;
  };

  auto pad = [](Eigen::MatrixXcRowMajor &data, size_t pad_width) {
    Eigen::MatrixXdRowMajor data_double = data.cast<double>();
    Eigen::MatrixXdRowMajor data_padded(data.rows() + 2 * pad_width,
                                        data.cols() + 2 * pad_width);
    data_padded.block((data_padded.rows() - data.rows() + 1) / 2,
                      (data_padded.cols() - data.cols() + 1) / 2, data.rows(),
                      data.cols()) = data_double;

    // Pad corners
    data_padded.topRightCorner(pad_width, pad_width)
        .setConstant(data_double(0, data.cols() - 1));
    data_padded.topLeftCorner(pad_width, pad_width)
        .setConstant(data_double(0, 0));
    data_padded.bottomLeftCorner(pad_width, pad_width)
        .setConstant(data_double(data.rows() - 1, 0));
    data_padded.bottomRightCorner(pad_width, pad_width)
        .setConstant(data_double(data.rows() - 1, data.cols() - 1));

    // Pad sides
    for (size_t i = 0; i < data.rows(); ++i) {
      data_padded.row(i + pad_width)
          .head(pad_width)
          .setConstant(data_double(i, 0));
      data_padded.row(i + pad_width)
          .tail(pad_width)
          .setConstant(data_double(i, data.cols() - 1));
    }
    for (size_t i = 0; i < data.cols(); ++i) {
      data_padded.col(i + pad_width)
          .head(pad_width)
          .setConstant(data_double(0, i));
      data_padded.col(i + pad_width)
          .tail(pad_width)
          .setConstant(data_double(data.rows() - 1, i));
    }

    return data_padded;
  };

  auto kernel = gaussian_kernel(kernel_size, sigma);
  auto data = pad(data_, (kernel_size - 1) / 2);

  Eigen::MatrixXcRowMajor output(data_.rows(), data_.cols());
  for (size_t y = 0, num_rows = data.rows(), num_cols = data.cols();
       y < num_cols - kernel_size + 1; ++y) {
    for (size_t x = 0; x < num_rows - kernel_size + 1; ++x) {
      output(x, y) = kernel
                         .cwiseProduct(data(Eigen::seq(x, x + kernel_size - 1),
                                            Eigen::seq(y, y + kernel_size - 1)))
                         .sum();
    }
  }

  return Image(output);
}

Eigen::MatrixXbRowMajor Image::Segment() {
  Eigen::MatrixXbRowMajor output = data_.cast<bool>();
  for (size_t i = 0, num_rows = output.rows(), num_cols = output.cols();
       i < num_rows; ++i) {
    for (size_t j = 0; j < num_cols; ++j) {
      output(i, j) = 1 ? data_(i, j) < 240 : 0;
    }
  }

  return output;
}

Eigen::Vector2d Image::CenterOfGravity() {
  if (!data_segmented_) {
    data_segmented_ = Segment();
  }
  int num_points = (*data_segmented_).count();

  Eigen::Vector2d mean_vector{0, 0};
  for (size_t i = 0, num_rows = (*data_segmented_).rows(),
              num_cols = (*data_segmented_).cols();
       i < num_rows; ++i) {
    for (size_t j = 0; j < num_cols; ++j) {
      mean_vector += Eigen::Vector2d{i, j} * (*data_segmented_)(i, j);
    }
  }
  return mean_vector / num_points;
}

Eigen::Matrix2d Image::PrincipalAxes() {
  if (!data_segmented_) {
    data_segmented_ = Segment();
  }
  int num_points = (*data_segmented_).count();
  auto mean_vector = CenterOfGravity();

  Eigen::Matrix2dRowMajor covariance_matrix{{0, 0}, {0, 0}};
  for (size_t i = 0, num_rows = (*data_segmented_).rows(),
              num_cols = (*data_segmented_).cols();
       i < num_rows; ++i) {
    for (size_t j = 0; j < num_cols; ++j) {
      if ((*data_segmented_)(i, j) == 0) {
        continue;
      }
      covariance_matrix += (Eigen::Vector2d{i, j} - mean_vector) *
                           (Eigen::Vector2d{i, j} - mean_vector).transpose();
    }
  }
  covariance_matrix = covariance_matrix / (num_points - 1);

  Eigen::SelfAdjointEigenSolver<Eigen::Matrix2dRowMajor> eigensolver;
  eigensolver.compute(covariance_matrix);
  Eigen::Vector2d eigen_values = eigensolver.eigenvalues().real().reverse();
  Eigen::Matrix2dRowMajor eigen_vectors =
      eigensolver.eigenvectors().real().rowwise().reverse();

  return eigen_vectors;
}

} // namespace image
} // namespace vision

bool operator==(const vision::image::Image &image1,
                const vision::image::Image &image2) {
  if (image1.data_ != image2.data_) {
    return false;
  }
  return true;
}
